runtime/dart_isolate.cc

```diff
@@ -803,10 +799,22 @@ DartIsolate::CreateDartVMAndEmbedderObjectPair(
 // |Dart_IsolateShutdownCallback|
 void DartIsolate::DartIsolateShutdownCallback(
     std::shared_ptr<DartIsolate>* embedder_isolate) {
+  // Terminate child Isolates
+  if ((*embedder_isolate)->terminate_child_isolates_) {
+    for (std::weak_ptr<DartIsolate> child_isolate :
+         (*embedder_isolate)->child_isolates_) {
+      if (!child_isolate.expired() && child_isolate.lock() != nullptr) {
```

Comment:
The `lock` will return an invalid shared ptr if the weak ptr has expired. So the first check is redundant.

See https://en.cppreference.com/w/cpp/memory/weak_ptr/lock

Reply:
I would move this to a `DartIsolate` member called `TerminateChildIsolatesIfNecessary` instead of doing multiple unrelated things in this one C callback. IMO, it makes the intent of the code clearer.